### PR TITLE
Big'N'Veiny pickups

### PR DESCRIPTION
**Pull request by husho**

### CODE REVIEW

**Comment by JayFoxRox** [comment]
I've skimmed over the code and it looks ok. I did not test it yet and I'm waiting for more information until I feel confident to give my final approval.

**Comment by husho** [comment]
> What's still missing to play the entire Big'n'Veiny mission?

It's almost playable, mission softlocks in the end, due to missing AI.

> How was this implemented? I assume the addresses are dumped from gta3.exe?

Yes

> Is there anything missing from this PR for the pickups?

If you are talking about only Big'n'Veiny pickups then I don't think so.

> Is there any specific behaviour which should be tested?

There is no special behaviour, except they only can be picked up in vehicle.

> I guess you can squash the "Minor fixes" commit into the first commit

Will do

File: rwengine/src/objects/PickupObject.hpp [hunk 1]
```diff
@@ -81,6 +85,34 @@ class PickupObject : public GameObject {
         return false;
     }
 
+    virtual bool isBigNVeinyPickup() const {
+        return false;
+    }
```
[the remark this answers is not in the document]
Comment (husho) [reply rwengine/src/objects/PickupObject.hpp]: GTAIII does this very poorly, IIRC, there is separate global pool for these pickups, and separate update call render call, and few specific creation and removal functions for opcodes.

[the remark this answers is not in the document]
Comment (husho) [reply rwengine/src/objects/PickupObject.hpp]: There is no, like, actual pickup type, which describes what this pickup is, in original game. We already have behavior specific type variable [same as in original game]. GTAIII just uses model index for that. For example if behavior type is PICKUP_ONCE or PICKUP_ONCE_TIMEOUT, etc it checks for model if it's health pickup model then give health. I guess we can add type enum, and ditch all these special classes maybe?

[the remark this answers is not in the document]
Comment (JayFoxRox) [reply rwengine/src/objects/PickupObject.hpp]: This is fine with me then (for now).

I guess we could have something like a `PickupPool` struct in the future which could be more similar to what GTA3 does.

But I feel this is such a small portion of the project that it's not worth time to redesign anything at this point. The solution is ugly, but it dos work.

File: rwgame/RWGame.cpp [hunk 2]
```diff
@@ -507,10 +507,13 @@ void RWGame::tick(float dt) {
             clockAccumulator -= 1.f;
         }
 
+        constexpr float timerClockRate = 1.f / 25.f;
```
Comment (JayFoxRox) [inline rwgame/RWGame.cpp]: I don't fully understand this change yet (I'm also not aware of the surrounding code).

Where does this `/ 25.f` come from - gta3.exe?
Does it have to do with FPS or what's the story behind it?

also nit for this commit: commit title is too long; should be reworded to not break into commit description

Comment (husho) [reply rwgame/RWGame.cpp]: > Does it have to do with FPS or what's the story behind it?

Basically, yes.

In GTAIII timers are updated every frame and framerate is locked to 30fps [if you have frame limiter on], there is no accumulators or anything. They may go out of sync depending on game's uptime.

File: rwengine/src/objects/PickupObject.hpp [hunk 3]
```diff
@@ -194,7 +229,134 @@ class MoneyPickup : public PickupObject {
         money = m;
     };
 
-    bool onCharacterTouch(CharacterObject* character) override;
+    bool onPlayerTouch() override;
+};
+
+const static std::array<glm::vec3, 106> bigNVeinyPickupsLocations = {
```
Comment (ShFil119) [inline rwengine/src/objects/PickupObject.hpp]: Can it be constexpr? It should be doable on almost every distro. (GLM 0.9.8.0 - 2016-09-11)

**Comment by danhedron** [comment]
LGTM

**Pull request merged**